lib/irb.rb

```diff
@@ -1114,23 +1111,46 @@ def build_statement(code)
       end
 
       code.force_encoding(@context.io.encoding)
-      command_or_alias, arg = code.split(/\s/, 2)
-      # Transform a non-identifier alias (@, $) or keywords (next, break)
-      command_name = @context.command_aliases[command_or_alias.to_sym]
-      command = command_name || command_or_alias
-      command_class = ExtendCommandBundle.load_command(command)
-
-      if command_class
-        Statement::Command.new(code, command, arg, command_class)
+      if (command, arg = parse_command(code))
+        command_class = ExtendCommandBundle.load_command(command)
+        Statement::Command.new(code, command_class, arg)
       else
         is_assignment_expression = @scanner.assignment_expression?(code, local_variables: @context.local_variables)
         Statement::Expression.new(code, is_assignment_expression)
       end
     end
 
-    def single_line_command?(code)
-      command = code.split(/\s/, 2).first
-      @context.symbol_alias?(command) || @context.transform_args?(command)
+    ASSIGN_OPERATORS = %w[= += -= *= /= %= **= &= |= &&= ||= ^= <<= >>=]
+    COMMAND_LIKE_ASSIGN_REGEXP = /\A[a-z_]\w* #{Regexp.union(ASSIGN_OPERATORS)}( |$)/
+
+    def parse_command(code)
+      command_name, arg = code.strip.split(/\s+/, 2)
+      return unless code.lines.size == 1 && command_name
+      return if COMMAND_LIKE_ASSIGN_REGEXP.match?(code)
+
+      arg ||= ''
+      command = command_name.to_sym
+      # Command aliases are always command. example: $, @
+      if (alias_name = @context.command_aliases[command])
+        return [alias_name, arg]
+      end
+
+      # Check visibility
+      local_variable = @context.local_variables.include?(command)
```

Comment:
While it's considerate to take local variables into account, I feel local variables should never override a command's invocation because:
- It's the behaviour of current IRB, Pry, and `debug`, so I'd expect most users to be aware of this limitation already.
- Under the current behaviour, users can still use `p <var>` to retrieve the variable when a conflict happens.
- However, under the this version users will NOT be able to call a command if a same-name local is present.

Reply:
I think the limitation is fine, but if we're going to fix #803, to be consistent, I think local variables should take precedence over methods.

```ruby
info = 1 # assignment is considered not a command in this pull request
=> 1
info + 1 # this will be a command call if we don't consider local variables
=> 2
```

Reply:
Yeah then it's probably better to leave the behaviour in #803 as is. I'll close it if you also agree.

Reply:
Although there's many things to consider, I want to keep #803 open because it will be an enhancement.